Add TypeScript tooling and configs

with-heart opened the pull request:
**What**:
This PR is an attempt to add TypeScript-related tooling and config required to complete the conversion from Flow to TypeScript. 

First step was getting `eslint` to run without any configuration errors (which @rjdestigter was running into in #2359), which seems to working for the most part. Also created a root `tsconfig.json` and tsconfig files for each package which extend from the root, since all of the existing were already duplicating the same config.

**Why**:
#2358

**How**:

- Created a root `tsconfig.json` which specifies the `compilerOptions` shared by all of the existing `tsconfig.json` files
	- Set `include` to `["packages", "scripts", "site", "test", "playgrounds"]`
	- Removed a few rules specifiers already enabled by `"strict": true`
- Created new and moved existing `tsconfig.json` files at root of each package
	- Specified `extends` for each to root `tsconfig.json`
	- Removed `compilerOptions` from all as they now inherit these from the root
	- Set `include` for each to `"src"`
- Configured `eslint` to support ts
	- Added `@typescript-eslint/parser` & `@typescript-eslint/eslint-config`
	- Switched `eslint-config-standard` to [`eslint-config-standard-with-typescript`](https://github.com/standard/eslint-config-standard-with-typescript)
	- Set `parserOptions.project` to root `tsconfig.json`
	- Disabled `no-use-before-define` and enabled `@typescript-eslint/no-use-before-define` as `warn`
- Added `@babel/preset-typescript` package and added to `presets` in `babel.config.js`

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->
<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation
- [ ] Tests
- [ ] Code complete
- [ ] Changeset N/A?
<!-- feel free to add additional comments -->

I haven't even attempted to build anything yet or actually convert files to ts. It seems like a lot of the base tooling- and config-related stuff probably needs to be dealt with first in order to support the conversion of actual packages.

I'd like to expand the checklist to include remaining items to handle, but I need some feedback and to explore a little more to determine what else there is to do.

rjdestigter commented:
Nice! I rebased (locally) my PR off your branch. It looks like there is a conflict between ESLint and Prettier when defining types and interfaces:

ESLint will complain about commas and semicolons: `Unexpected separator (;).eslint@typescript-eslint/member-delimiter-style`

While prettier seems to think we have bad syntax when there is neither: `(property) SerializedStyles.styles: string`

![image](https://user-images.githubusercontent.com/14314519/116272734-6210a100-a73e-11eb-821c-25f33e72cb4f.png)

Andarist commented on package.json:
```diff
@@ -51,7 +53,7 @@
   "eslintConfig": {
     "extends": [
-      "standard",
+      "standard-with-typescript",
```
I think we can just drop the standard plugin since we are using Prettier.

Andarist commented on package.json:
```diff
@@ -27,6 +27,8 @@
     "release": "yarn build && changeset publish"
   },
   "resolutions": {
+    "**/@typescript-eslint/eslint-plugin": "^4.22.0",
+    "**/@typescript-eslint/parser": "^4.22.0",
```
any particular reason for those resolutions here? 🤔

with-heart replied:
`gatsby` has a dependency on `^2.x` for each, and it was causing linting to fail. I can investigate more to see if there's a better way, this was just the quick and easy solution I Google'd to avoid the conflict.

Andarist replied:
Hm, why does Gatsby depend on this anyway? It shouldn't be concerned with linting by default 🤔

with-heart replied:
I removed the `resolutions` and `node_modules` and reinstalled and it seems to be working fine now. I'm not sure what happened there exactly, though `gatsby` fixed that issue in February so maybe I had an outdated version installed.

Andarist commented on babel.config.js:
```diff
@@ -10,7 +10,7 @@ let isTestFile = filename =>
   api.cache(true)
   return {
-    presets: ['babel-preset-emotion-dev'],
+    presets: ['babel-preset-emotion-dev', '@babel/preset-typescript'],
```
isn't this conflicting with the flow plugin that is used at the same time?

with-heart replied:
Probably but I haven't tried to build anything yet. Not sure how to coordinate that part exactly if do conflict. Just remove `flow` and assume we won't merge all of these branches till everything is converted? 

Added because `preconstruct` uses it.

mitchellhamilton replied:
We should use https://babeljs.io/docs/en/options#overrides to make run @babel/preset-typescript on `.ts` and `.tsx` and `@babel/preset-flow` on `.js`. Ideally we should merge these incrementally.

Andarist replied:
note that `@babel/preset-typescript` already has `overrides` in its implementation: https://github.com/babel/babel/blob/672a58660f0b15691c44582f1f3fdcdac0fa0d2f/packages/babel-preset-typescript/src/index.js#L35 . I'm not completely sure how well this composes with other stuff though.

with-heart replied:
It doesn't appear to conflict. `build` works fine whether the preset is listed or not.

Andarist commented on package.json:
```diff
@@ -81,7 +86,9 @@
       "standard/computed-property-even-spacing": 0,
-      "@emotion/pkg-renaming": 2
+      "@emotion/pkg-renaming": 2,
+      "no-use-before-define": 0,
+      "@typescript-eslint/no-use-before-define": 1
```
interesting - isn't it one of the goals of TypeScript ESLint project to hide this from consumers? 🤔 I mean - don't they automatically switch to TS-oriented variants of plugins like this?

with-heart replied:
I think we might need to enable the `recommended` rules from `@typescript-eslint/eslint-config` for that. I didn't want to enable it since I wasn't sure how well it'd play together with `standard`. Since we're dropping that, I'll enable the recommended rules and see how that goes.

with-heart replied:
Yeap, `extends:@typescript-eslint/eslint-recommended" removed the need for this.

Andarist commented on packages/babel-plugin/tsconfig.json:
```diff
@@ -0,0 +1,4 @@
+{
```
For the time being, I would leave the existing tsconfig.json files as they are targeting **test** directories and not really the source files. Keeping them together with type tests might actually be a requirement of DTslint or something - but this has to be investigated later.

As to the project itself - I think a single `tsconfig.json` in the root will work great for us. This is how TS is set up in Changesets repo.

with-heart replied:
Good point. I'll put those back!

with-heart commented:
@rjdestigter Yeah, I noticed that as well! I think it has to do with `prettier` being unsure which parser to use. Need to investigate more.

with-heart commented:
I made a few changes that moved this forward:

- removed `standard` plugins
- added `import`, `react`, and `@typescript-eslint` as `eslint` plugins
- added an `eslint` files override for `**/packages/**/*.tsx?` that sets the `prettier` parser to `typescript`

This puts us at 38 errors and 61 warnings. The warnings are all for `@typescript-eslint/no-use-before-define`, and the errors are split between the parser being unable to deal with `flow` and `parserOptions.project` confusion for root `*.js` config files and `**/__mocks__/*.js` files. Not sure of the best way to resolve that last issue.

with-heart commented:
Hoping to take a look at the review comments later today.

Andarist commented on package.json:
```diff
@@ -60,6 +60,8 @@
       "prettier",
       "flowtype",
       "@emotion",
+      "import",
```
I wonder - was it required for this PR? what problem does it solve here?

with-heart replied:
It was needed for this override: https://github.com/emotion-js/emotion/blob/main/package.json#L99-L107

But [the docs for that rule state](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-commonjs.md) that it's meant to be temporary for transitioning to pure ES6 modules, so I think it's fine to remove this plugin and the override.

Andarist commented on package.json:
```diff
@@ -110,6 +110,20 @@
+          "prettier/prettier": [
+            "error",
+            {
+              "parser": "typescript"
```
since we are going to use Babel for parsing TS maybe it would be better to use `babel-ts` here instead?

with-heart replied:
That sounds fine! We'd need to upgrade `prettier` in order to use it though. Is that fine to include?

Andarist replied:
ye, go for it

Andarist replied:
actually - maybe leave that off for a followup PR

with-heart commented:
I made a few more changes this morning. `build` works fine and linting now results in 32 errors related to `@typescript-eslint/eslint-parser` being unable to parse `flow`.

If I add an `overrides` section for `**/packages/**/*.jsx?" with `"parser": "babel-eslint"`, this drops to only 5 linting errors related to `flow`. I don't know enough about `flow` to know how to resolve those though, and I'm not sure if we'd want to include that override temporarily while the conversion to ts happens.

Thoughts?

michaeljaltamirano commented on packages/babel-preset-css-prop/tsconfig.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Can't leave a comment on the diff but `packages/babel-plugin-jsx-pragmatic/tsconfig.json` being empty looks to be the cause for [the Flow CI issue](https://github.com/emotion-js/emotion/pull/2360/checks?check_run_id=2504489068).

JoshuaKGoldberg commented on tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+  },
+  "include": ["packages", "scripts", "site", "test", "playgrounds"],
+  "exclude": ["node_modules"]
+}
```
Well never mind then 😄

JoshuaKGoldberg commented on tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "lib": ["es6", "dom"],
+    "module": "commonjs",
+    "noEmit": true,
+    "strict": true,
```
Just a suggestion: I get a lot of errors around conflicting types, generally around react-native and the `dom` types. 

```suggestion
    "skipDefaultLibCheck": true,
    "strict": true,
```

JoshuaKGoldberg commented on tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "target": "es5"
+  },
+  "include": ["packages", "scripts", "site", "test", "playgrounds"],
+  "exclude": ["node_modules"]
```
I've found myself doing this in my own integration branch to get TypeScript to not complain about all the `$ExpectError` dts tests:

```suggestion
  "exclude": ["node_modules", "packages/*/types/asdf*"]
```

JoshuaKGoldberg replied:
I guess, more broadly, what is the strategy for `.d.ts` generation? Should a build step like `build:types` be added that runs `tsc --declaration` into `types/` folders?

Andarist replied:
> I guess, more broadly, what is the strategy for .d.ts generation? Should a build step like build:types be added that runs tsc --declaration into types/ folders?

This one is covered by Preconstruct which we are using to build all packages. You can check out example output containing TS declarations here:
https://unpkg.com/browse/@changesets/cli@2.14.0/dist/

JoshuaKGoldberg commented on package.json:
```diff
@@ -264,6 +270,7 @@
     "react-test-renderer": "16.8.6",
     "svg-tag-names": "^1.1.1",
     "through": "^2.3.8",
+    "typescript": "^4.2.4",
```
I think `prettier` is going to need to be upgraded as well. `prettier@1.14.3` is pretty old and crashes on new TypeScript syntax such as `import type ...`.

```
$ C:\Code\emotion\node_modules\.bin\prettier packages\utils\src\index.ts
[error] packages\utils\src\index.ts: SyntaxError: '=' expected. (1:13)
[error] > 1 | import type { RegisteredCache, EmotionCache, SerializedStyles } from './types'
[error]     |             ^
```

michaeljaltamirano replied:
Could the syntax omit `type` until prettier is upgraded? Or are there other blockers present too.

Andarist replied:
We could just upgrade Prettier - probably best to do it in a separate, dedicated, PR after removing flow types.

JoshuaKGoldberg replied:
https://github.com/emotion-js/emotion/pull/2387 for reference.

My two cents: a Prettier upgrade is generally a nice thing and doesn't relate to the rest of the TypeScript stuff. Could it come in separately, then get brought into the `ts-migration` as a merge from master?

JoshuaKGoldberg replied:
Merged 🎊

with-heart commented:
Welp, I've updated the branch but I was being mindless and rebased. My bad for throwing a wrench into things, wasn't really thinking.

JoshuaKGoldberg commented:
This all looks good either way 😄 , I'll get my branch updated hopefully end of this week!

JoshuaKGoldberg commented:
This all looks good either way 😄 , I'll get my branch updated hopefully end of this week!

@with-heart, other than updating https://github.com/JoshuaKGoldberg/emotion/pull/1, is there anything you'd like me to do to help? I think once that's updated it would be ready to be merged in here?

Andarist commented on packages/babel-plugin-jsx-pragmatic/tsconfig.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Could u try removing those per-directory tsconfigs? Do they change anything in how things work now? In Changesets, this is all we have: https://github.com/atlassian/changesets/blob/c426035565cfac518238c8bf32f3c496c66c0657/tsconfig.json

Andarist commented:
@with-heart friendly 🏓 would you mind taking a look at what's up with CI?

with-heart commented:
I fixed the issue with the `eslint` parser for `prettier` and removed the per-package `tsconfig.json` files. Are you able to re-reun the workflows for me @Andarist?

Andarist commented:
@with-heart thank you for this! ❤️
